File: backend/src/main/java/com/team01/backend/domain/post/repository/PostLikeRepository.java
package com.team01.backend.domain.post.repository;

import com.team01.backend.domain.post.entity.PostLike;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface PostLikeRepository extends JpaRepository<PostLike, Long> {

    @Query("SELECT pl FROM PostLike pl WHERE pl.user.id = :userId AND pl.post.id = :postId")
    Optional<PostLike> findByUserIdAndPostId(@Param("userId") Long userId, @Param("postId") Long postId);

    @Query("SELECT COUNT(pl) FROM PostLike pl WHERE pl.post.id = :postId")
    int countByPostId(@Param("postId") Long postId);

    @Modifying
    @Query("DELETE FROM PostLike pl WHERE pl.user.id = :userId AND pl.post.id = :postId")
    int deleteByUserIdAndPostId(@Param("userId") Long userId, @Param("postId") Long postId);

    @Modifying
    @Query(value = "INSERT INTO POST_LIKES (USER_ID, POST_ID) SELECT :userId, :postId WHERE NOT EXISTS (SELECT 1 FROM POST_LIKES WHERE USER_ID = :userId AND POST_ID = :postId)", nativeQuery = true)
    int tryInsert(@Param("userId") Long userId, @Param("postId") Long postId);

    List<PostLike> findByPost_Id(Long postId);
}

File: backend/src/main/java/com/team01/backend/domain/post/repository/PostLikeRepository.kt
package com.team01.backend.domain.post.repository

import com.team01.backend.domain.post.entity.Post
import com.team01.backend.domain.post.entity.PostLike
import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface PostLikeRepository : JpaRepository<PostLike, Long> {

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT p FROM Post p JOIN FETCH p.board WHERE p.id = :id")
    fun findPostByIdForUpdate(@Param("id") id: Long): Post?

    fun findByPost_IdAndUser_Id(postId: Long, userId: Long?): PostLike?

    fun deleteByPost_IdAndUser_Id(postId: Long, userId: Long?): Int

    @Query("SELECT COUNT(pl) FROM PostLike pl WHERE pl.post.id = :postId")
    fun countByPostId(@Param("postId") postId: Long): Int

    fun findByPostId(postId: Long): List<PostLike>

}